flashinfer/aot.py

```diff
@@ -0,0 +1,504 @@
+from .cascade import gen_cascade_module
+from .custom_all_reduce import gen_comm_module
+from .gemm import gen_gemm_module, gen_gemm_sm90_module, gen_gemm_sm100_module
+from .jit import JitSpec, build_jit_specs
+from .jit import env as jit_env
+from .jit import (
+    gen_batch_decode_module,
+    gen_batch_mla_module,
+    gen_batch_prefill_module,
+    gen_fmha_cutlass_sm100a_module,
+    gen_single_decode_module,
+    gen_single_prefill_module,
+)
+from .norm import gen_norm_module
+from .page import gen_page_module
+from .quantization import gen_quantization_module
+from .rope import gen_rope_module
+from .sampling import gen_sampling_module
+
+
+def gen_gqa_fa2(
+    dtype_qo: torch.dtype,
+    dtype_kv: torch.dtype,
+    head_dim_qk: int,
+    head_dim_vo: int,
+    use_sliding_window: bool,
+    use_logits_soft_cap: bool,
+) -> List[JitSpec]:
+    if dtype_qo.itemsize == dtype_kv.itemsize and dtype_qo != dtype_kv:
+        return []
+    return [
+        gen_single_prefill_module(
+            backend="fa2",
+            dtype_q=dtype_qo,
+            dtype_kv=dtype_kv,
+            dtype_o=dtype_qo,
+            head_dim_qk=head_dim_qk,
+            head_dim_vo=head_dim_vo,
+            pos_encoding_mode=0,
+            use_sliding_window=use_sliding_window,
+            use_logits_soft_cap=use_logits_soft_cap,
+            use_fp16_qk_reduction=False,
+        ),
+        gen_batch_prefill_module(
+            backend="fa2",
+            dtype_q=dtype_qo,
+            dtype_kv=dtype_kv,
+            dtype_o=dtype_qo,
+            dtype_idx=torch.int32,
+            head_dim_qk=head_dim_qk,
+            head_dim_vo=head_dim_vo,
+            pos_encoding_mode=0,
+            use_sliding_window=use_sliding_window,
+            use_logits_soft_cap=use_logits_soft_cap,
+            use_fp16_qk_reduction=False,
+        ),
+        gen_single_decode_module(
+            dtype_q=dtype_qo,
+            dtype_kv=dtype_kv,
+            dtype_o=dtype_qo,
+            head_dim_qk=head_dim_qk,
+            head_dim_vo=head_dim_vo,
+            pos_encoding_mode=0,
+            use_sliding_window=use_sliding_window,
+            use_logits_soft_cap=use_logits_soft_cap,
+        ),
+        gen_batch_decode_module(
+            dtype_q=dtype_qo,
+            dtype_kv=dtype_kv,
+            dtype_o=dtype_qo,
+            dtype_idx=torch.int32,
+            head_dim_qk=head_dim_qk,
+            head_dim_vo=head_dim_vo,
+            pos_encoding_mode=0,
+            use_sliding_window=use_sliding_window,
+            use_logits_soft_cap=use_logits_soft_cap,
+        ),
+    ]
+
+
+def gen_gqa_fa3(
+    dtype_qo: torch.dtype,
+    dtype_kv: torch.dtype,
+    head_dim_qk: int,
+    head_dim_vo: int,
+    use_sliding_window: bool,
+    use_logits_soft_cap: bool,
+) -> List[JitSpec]:
+    if dtype_qo.itemsize == dtype_kv.itemsize and dtype_qo != dtype_kv:
+        return []
+    if dtype_kv.itemsize == 1:
+        # fp8 kv not supported in FA3
```

Comment:
it's supported in https://github.com/flashinfer-ai/flashinfer/pull/1033

Reply:
Yes. That's my impression. But I failed to compile. Let me try again.

Reply:
Seems there are compilation bugs with certain head dims for fp8 fa3. Let's ignore them and fix in later PRs.